### ISSUE TITLE
Bring back JDBI support

### ISSUE THREAD

**Issue by augustotravillio** [immutables/immutables#97]
JDBI integration was available in 1.1.x (https://github.com/immutables/immutables/tree/fix1.1/common/src/org/immutables/common/jdbi)
It was dropped in 2.0. As far as I can understand it was based on old JSON marshaling infrastructure as a binding engine. Could we restore this integration, maybe based on new gson bindings? It should probably be packaged as separate module like `org.immutables:jdbi`.

**Comment by jdmwood2** [comment]
I managed to backport JDBI support for 2.0.

You can see the Gist here:

https://gist.github.com/jdmwood2/4c333e27bfe689cb3b07

I've basically tried to keep it as close to the original files as possible.

This should basically be a drop in replacement.

The only caveat is that I've added support for Joda DateTime and I've only really tried this out with MySQL.

If you don't need dates, or you don't use DateTime you might have to fiddle a bit.

Enjoy!

**Comment by elucash** [comment]
Wow, this is great, thanks! I think we will incorporate that in new jdbi module

**Comment by jdmwood2** [comment]
No problem! Glad to help.

**Comment by brianm** [comment]
Hi hi, I know jdbi pretty well, and am learning my way around immutables -- does this need any help to get moving again? I'd be thrilled to nudge it along (mostly as I want to use it!).

**Comment by brianm** [comment]
This looks like it roundtrips through jackson, which is not horrible, but I'd love to find a way to generate exact mappers for values rather than do it dynamically. Can the code generation cycle for values be hooked into in any way?

**Comment by elucash** [comment]
Hi! I can only encourage contribution to this! Having JDBI support is really nice to have feature, but  we could not allocate enough time and interest to put into quality JDBI support.
Currently we are working on extensibility, but so far it is not based on extending annotation processor but prov),. We could easily introduce extension point in a pipeline as necessary. Another possibility is creating a separate annotation processor and use some utility methods/pieces of infrastructure to get more elaborate models that immutables use .
But as a starting point I would strongly suggest just to fork Immutables and create separate generator and add it to https://github.com/immutables/immutables/blob/master/value-processor/src/org/immutables/value/processor/Processor.java . Unfortunately template syntax and apis are undocumented, but a lot of stuff could be easily picked up. There's possible to use any tool to produce the source code and glue things together. Let me know of specific question you might have.

Also, just to bring some context, as we are generating TypeAdapters for Gson, there was also idea to use Gson as a basis for converting an object in a set of fields. And Gson.generator could serve as example of per-attribute, per-type code generation: https://github.com/immutables/immutables/blob/master/value-processor/src/org/immutables/value/processor/Gsons.generator

**Comment by brianm** [comment]
Sweet, I'll take a look!

**Comment by Gaff** [comment]
Possibly of little use to this thread but I also knocked up a solution to this. It's a DIY annotaion processor that sidesteps the immutables stuff though and simply spits out a ResultSetMapper. I'm happy to chip in what we have, however the biggest issue is that I couldn't work out how to get the immutables template stuff to work, so just used moustache instead.

Still if the model code is useful I'm happy to offer it up. Plus can help with testing etc

**Comment by elucash** [comment]
@Gaff I guess the most value is in the solution. This would be very awesome if you can share it! Re-coding a template or plugging in moustache renderer might be half-an-hour work and is negligible compared to the time it took to came up with models, structure of generated classes and to test it.

**Comment by Gaff** [comment]
Well ok, see here:  https://gist.github.com/Gaff/f65c39ec1bed36d6b9d3da425ca959a6

The details aren't exactly rocket science. The rough plot for us is that we need a ```ResultSetMapper``` instance for each immutable type (We also need a binder, not included here as it's a simpler problem). I've got a moustache template that looks like psuedo-java of what the mapper should look like that delegates some work to a helper class. Behind it we have a simplistic model that works out the concrete details, and an annotation processor plus annotations to glue the whole thing together. See my comment at the bottom of the gist for a sample input / output.

Note that my solution is rough around the edges, for example we're using the jackson object building when we probably ought to be using the JDBI version (argument factory?). Error handling is poor, and no doubt there are many many edge cases I've not covered. I also generate some SQL insert statements - I realise this is highly opinionated, but for many simple cases it removes a lot of error prone statement carving.

Anyhow apologies for the wartiness of the whole thing. It's there if it's useful. If not feel free to ignore.

**Comment by brianm** [comment]
Given the huge flexibility in naming of builders, is there a deterministic way to build immutables? Am looking at code genning mappers for each, a la @elucash 's approach -- but would like to make it work in the face of naming schemes, etc.

**Comment by stevenschlansker** [comment]
hi @elucash @Gaff @brianm jdbi project checking in again, there's renewed interest in this feature.
I do not want it to rely on Jackson integration; as both projects only optionally use it, we should be able to do without.  The two obvious approaches I see are 

1) standardized metadata (property name, reader, writer) usable by external engines.  This has the appeal of doing it once and anyone who wants to answer questions about the Immutables types just has to read its metadata.  Maybe this takes the form of an annotation on generated classes that points to a descriptor.

2) Code generate jdbi integration.  I've been learning the templating engine enough to be dangerous, although setting up a whole new integration like Gson is still not obvious to me.  This would likely involve a new `Jdbi.generator` that builds appropriate `RowMapper` and `ArgumentFactory` types.

I'd probably use the above references as a starting point but anything that wants to get merged into mainline will have to be much more general.

Any thoughts from any of the contributors?  Thanks!

**Comment by stevenschlansker** [comment]
Also, if we do this, you get to close one of your oldest issues!  Wouldn't that be a happy day?

**Comment by stevenschlansker** [comment]
A lightweight (1) could actually be even simpler -- we could annotate each generated method with e.g. `@ImmutableGetter` `@ImmutableBuilderFrom` (an annotation to describe what the method does) and guess the rest from the types.  Really, the only sticking point is that we need a more durable way to detect available properties if someone customizes the generated method names in a crazy way.

**Comment by elucash** [comment]
fully agree on the set of options, very glad that we're thinking likewise. Some discussions about this scattered across comments under different issues, but you summarized it well. I would also very like start working on it and, of course, don't have much time, so help is definitely needed. The metadata would be overall more useful for a range of different integrations, but we need to came up with high quality API for elements etc. Generated mappers for JDBI (I guess v3 is relevant nowadays) would not require such a high quality API as the thing would just be an implementation generated against JDBI's API and should be a lot easier than Gson.

**Comment by Gaff** [comment]
Still interested in any solutions to this. We're still using the solution above, we replaced Mustache with JavaPoet to do the code generation but the idea is the same - basically annotate the immutable class with some clues about how to map the SQL columns, then generate a `ResultSetMapper`. Happy to offer-up further details if anyone cares.

**Comment by elucash** [comment]
@Gaff really cool! do you have the upgraded version released? I wonder what are the main pain-points which Immutables can address so that the solution like yours could be just used. Or what are major lessons learned that we need to understand if we're to replicate the solution

**Comment by Gaff** [comment]
The code is probably rather specific to our use cases, I'd need to check before releasing it.  Checkout my gist earlier in this thread to get an idea, if you think it would be useful to see more of that code let me know and I'll see what I can do.

**Comment by elucash** [comment]
yep, I've looked at the gist previously and today, quite a lot of good work. The most time-consuming part in such solutions, I guess, is to get type handling/conversions right. I can get to it someday, but I have many projects in parallel, so barely have any capacity to hold this in my head, if any of you ( @Gaff  @stevenschlansker ) want to try to move this, I can help by laying out some groundwork / structure, and/or answering questions.

**Comment by stevenschlansker** [comment]
Luckily jdbi 3 has pretty robust type handling -- that's one of the things we worked a lot on!  As long as it can figure out the property name and method to read / write, we can reuse our JavaBean support.  I am hacking on a prototype now.

**Comment by brianm** [comment]
FWIW, having a consistent programmatic interface available for ANY immutable (regardless of style) would be amazing. It could even rely on service locator loading a custom implementation of the "reflection" api for the immutable. This is one of our biggest pain points in writing things that work with immutables, despite asking everyone to use the same styles, they don't, and it would be nice to not care about styles from a programmatic POV

**Comment by brianm** [comment]
@elucash that would be awesome -- it would solve my biggest headache with immutables (that I like code generation in general, for more than just the data objects) :-)

**Comment by Gaff** [comment]
@elucash - Here's an updated Gist showing how we build the ResultSetMapper using JavaPoet:

https://gist.github.com/Gaff/ab8c45163e95c98b49f25a6c0058cbe5 - I'll try and tidy up everything into a package I can publish. It's rather intertwined with other things.

We also have code that does the `@bind`ing via reflection.

**Comment by stevenschlansker** [comment]
I've started basic work here, and have a simple test case passing!
https://github.com/jdbi/jdbi/pull/1282
Right now I just assume default naming convention and style.  None of the annotations are retained at runtime, so I cannot discover anything at all about the class shape.

**Comment by stevenschlansker** [comment]
We are now running the new Immutables support in production.  Aside from a couple bugs and corner cases, things work fairly well.  Anyone else have a chance to give it a try yet?
